File: novelos-longform/novelos-app/src/lib/chapterOutline.ts
export interface PlotPoint {
  order: number;
  description: string;
  characters_involved: string[];
  estimated_words: number;
}

export interface ChapterOutlineSections {
  开篇场景: string;
  情节点列表: PlotPoint[];
  角色出场: string[];
  关键对话: string[];
  转折点: string;
  章末状态: string;
  总字数估算?: number;
}

function stripJsonFence(content: string): string {
  const trimmed = content.trim();
  const fenced = trimmed.match(/^```(?:json|JSON)?\s*([\s\S]*?)\s*```$/);
  return (fenced?.[1] ?? trimmed).replace(/^\uFEFF/, "").trim();
}

function parseJsonLike(content: string): unknown | null {
  const cleaned = stripJsonFence(content);
  try {
    return JSON.parse(cleaned);
  } catch {
    const start = cleaned.indexOf("{");
    const end = cleaned.lastIndexOf("}");
    if (start < 0 || end <= start) return null;
    try {
      return JSON.parse(cleaned.slice(start, end + 1));
    } catch {
      return null;
    }
  }
}

function asRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== "object" || Array.isArray(value)) return null;
  return value as Record<string, unknown>;
}

function asNumber(value: unknown): number {
  const n = Number(value);
  return Number.isFinite(n) ? n : 0;
}

function stringifyInline(value: unknown): string {
  if (typeof value === "string") return value;
  if (typeof value === "number" || typeof value === "boolean") return String(value);
  if (Array.isArray(value)) return value.map(stringifyInline).filter(Boolean).join("、");
  const record = asRecord(value);
  if (!record) return "";
  return Object.values(record).map(stringifyInline).filter(Boolean).join("；");
}

function normalizeStringList(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value
    .map(stringifyInline)
    .map((s) => s.trim())
    .filter(Boolean);
}

function normalizeCharacters(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value
    .map((item) => {
      if (typeof item === "string") return item;
      const record = asRecord(item);
      if (!record) return "";
      const name = stringifyInline(record.name);
      const role = stringifyInline(record.role_in_chapter ?? record.role);
      const action = stringifyInline(record.key_action ?? record.action);
      return [name, role, action].filter(Boolean).join("：");
    })
    .map((s) => s.trim())
    .filter(Boolean);
}

function normalizeDialogues(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value
    .map((item) => {
      if (typeof item === "string") return item;
      const record = asRecord(item);
      if (!record) return "";
      const participants = stringifyInline(record.participants);
      const purpose = stringifyInline(record.purpose);
      const setting = stringifyInline(record.setting);
      return [participants && `【${participants}】`, purpose, setting && `场景：${setting}`]
        .filter(Boolean)
        .join(" ");
    })
    .map((s) => s.trim())
    .filter(Boolean);
}

function normalizePlotPoints(value: unknown): PlotPoint[] {
  if (!Array.isArray(value)) return [];
  return value
    .map((item, index) => {
      if (typeof item === "string") {
        return {
          order: index + 1,
          description: item,
          characters_involved: [],
          estimated_words: 0,
        };
      }
      const record = asRecord(item);
      if (!record) return null;
      const description = stringifyInline(record.description ?? record.desc ?? record.content);
      if (!description) return null;
      return {
        order: asNumber(record.order) || index + 1,
        description,
        characters_involved: normalizeStringList(
          record.characters_involved ?? record.characters ?? record.roles,
        ),
        estimated_words: asNumber(record.estimated_words ?? record.words),
      };
    })
    .filter((point): point is PlotPoint => Boolean(point));
}

function parsePlotPointsFromText(text: string): PlotPoint[] {
  if (!text.trim()) return [];
  return text
    .split("\n")
    .map((line) => line.trim())
    .filter(Boolean)
    .map((line, idx) => ({
      order: idx + 1,
      description: line.replace(/^[-*\d.]+\s*/, "").trim(),
      characters_involved: [],
      estimated_words: 0,
    }))
    .filter((point) => Boolean(point.description));
}

function parseListFromText(text: string): string[] {
  if (!text.trim()) return [];
  return text
    .split("\n")
    .map((l) => l.replace(/^[-*\d.]+\s*/, "").trim())
    .filter(Boolean);
}

function normalizeOutlineData(data: Record<string, unknown>): ChapterOutlineSections {
  return {
    开篇场景: stringifyInline(data.开篇场景 ?? data.opening_scene),
    情节点列表: normalizePlotPoints(data.情节点列表 ?? data.plot_points),
    角色出场: normalizeCharacters(
      data.角色出场 ?? data.character_appearances ?? data.characters_appearing,
    ),
    关键对话: normalizeDialogues(data.关键对话 ?? data.key_dialogues),
    转折点: stringifyInline(data.转折点 ?? data.turning_point),
    章末状态: stringifyInline(data.章末状态 ?? data.ending_state ?? data.chapter_end_state),
    总字数估算: asNumber(data.total_estimated_words) || undefined,
  };
}

export function parseOutlineContent(content: string): ChapterOutlineSections | null {
  const parsed = parseJsonLike(content);
  const parsedRecord = asRecord(parsed);
  const resultRecord = asRecord(parsedRecord?.result) ?? parsedRecord;

  if (resultRecord) {
    return normalizeOutlineData(resultRecord);
  }

  const sections: Record<string, string> = {};
  const sectionNames = ["开篇场景", "情节点列表", "角色出场", "关键对话", "转折点", "章末状态"];
  let currentSection = "";
  const lines = content.split("\n");

  for (const line of lines) {
    const match = line.match(/^#{1,3}\s*(.+)/) || line.match(/^【(.+)】/);
    if (match) {
      const name = match[1].trim();
      const found = sectionNames.find((s) => name.includes(s));
      if (found) {
        currentSection = found;
        sections[currentSection] = "";
        continue;
      }
    }
    if (currentSection) {
      sections[currentSection] += (sections[currentSection] ? "\n" : "") + line;
    }
  }

  if (Object.keys(sections).length === 0) return null;

  return {
    开篇场景: sections["开篇场景"] || "",
    情节点列表: parsePlotPointsFromText(sections["情节点列表"] || ""),
    角色出场: parseListFromText(sections["角色出场"] || ""),
    关键对话: parseListFromText(sections["关键对话"] || ""),
    转折点: sections["转折点"] || "",
    章末状态: sections["章末状态"] || "",
  };
}

File: novelos-longform/novelos-app/src-tauri/src/agents/chapter_outline.rs
pub const SYSTEM: &str = r#"你是一个专业的章节大纲规划专家。你需要根据章节任务卡和上下文，生成详细的章节大纲。

章节大纲结构：
1. **开篇场景**: 章节开头的情境设定（谁、在哪里、在做什么）
2. **情节推进**: 3-5个关键情节点，按顺序排列
3. **角色互动**: 本章出现的角色及其互动关系
4. **关键对话**: 1-2段关键对话的场景和目的
5. **转折点**: 本章中发生的转折或揭示
6. **章末状态**: 本章结束时的状态变化

写作原则：
- 节奏要有张有弛，避免平铺直叙
- 对话要推动剧情，避免空洞寒暄
- 描写的分配要合理（环境/动作/心理/对话）
- 严格遵守正典规则
- 确保人物言行符合SOUL设定
- 大纲节奏要贴合题材模板中的卷节奏、爽点类型和禁忌规则
- 全书大纲与书籍设定是最高优先级约束，开篇场景、角色、冲突和章末状态不得偏离既定书籍信息

输出格式：只输出一个 JSON 对象，不要添加 Markdown 代码块、说明文字或前后缀。
{
  "chapter_number": 1,
  "opening_scene": "开篇情境描述",
  "plot_points": [
{"order": 1, "description": "情节描述", "characters_involved": ["角色"], "estimated_words": 500},
{"order": 2, "description": "...", "characters_involved": ["角色"], "estimated_words": 600}
  ],
  "character_appearances": [
{"name": "角色名", "role_in_chapter": "本章作用", "key_action": "核心行为"}
  ],
  "key_dialogues": [
{"participants": ["A", "B"], "purpose": "对话目的", "setting": "对话场景"}
  ],
  "turning_point": "本章转折点描述",
  "ending_state": "章末状态变化",
  "total_estimated_words": 3000
}"#;

pub const USER_TEMPLATE: &str = "请为以下章节生成详细大纲：\n\n章节任务卡：\n{task_card}\n\n全书大纲与书籍设定（必须优先遵守）：\n{book_outline_context}\n\n题材模板约束：\n{genre_template}\n\n卷纲上下文：\n{volume_context}\n\n前5章大纲记忆（避免重复）：\n{recent_chapter_outlines}\n\n适用正典规则：\n{canon_rules}\n\n角色SOUL参考：\n{soul_refs}";
